### PR TITLE
Table entries

### PR DESCRIPTION
**Pull request by KinanBab**
Parse table entries and embed them in IR.

### CODE REVIEW

File: p4_symbolic/ir/ir.proto [hunk 1]
```diff
@@ -93,9 +93,22 @@ message TableImplementation {
...
+  p4_symbolic.bmv2.TableMatchType match_type = 1;
   // Table type with respect to action selection.
-  p4_symbolic.bmv2.ActionSelectorType action_selector_type = 6;
+  p4_symbolic.bmv2.ActionSelectorType action_selector_type = 2;
+  // Concrete table entries.
+  repeated TableEntry entries = 3;
+}
+
+// Table entry parsed from P4RT/simple_switch_CLI.
+message TableEntry {
```
Comment (stefanheule) [inline p4_symbolic/ir/ir.proto]: Is there a reason we are not using the p4.v1.TableEntry instead of this?

**Comment by KinanBab** [comment]
Note: there are protobuf definitions for what a table entry looks like in pdpi https://github.com/google/p4-pdpi/blob/master/p4_pdpi/ir.proto.

We can reuse that, or reuse the one in p4runtime https://github.com/p4lang/p4runtime/blob/f324cba9b213c617de94ffeb0e1bd65d3aa32e19/proto/p4/v1/p4runtime.proto.

I am leaning more towards the p4runtime one, since it can parse all the data formats for us (e.g. ipv4, ipv6) as bit strings, which we can use directly. Another consideration is which can parse the entries out of the box for us.

I am planing to do this switch later on, for now I want to use this current code as a shortcut to get the end-end prototype working. When I do this switch, this will essentially change the proto definition from this PR to use one of the two options above, and will likely change table_entries.cc to be more like pdpi_driver.cc (i.e. a wrapper around calls to the library we decide on to parse the data for us). The rest of the code will remain similar.

File: p4_symbolic/ir/ir.proto (hunk 1 above)
Comment (KinanBab) [reply p4_symbolic/ir/ir.proto]: You beat me to this, I was writing a comment below about this. Please see the comment.

File: p4_symbolic/ir/table_entries.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include "p4_pdpi/utils/status_utils.h"
+#include "p4_symbolic/ir/ir.pb.h"
+
+namespace p4_symbolic {
+namespace ir {
+
+using TableEntries = std::vector<std::pair<std::string, TableEntry>>;
+
+// Parses entries read from entries_path, and fills them in given ir in place.
+pdpi::StatusOr<TableEntries> ParseAndFillEntries(const char *entries_path);
```
Comment (konne88) [inline p4_symbolic/ir/table_entries.h]: Use string instead of char*

File: p4_symbolic/main.cc [hunk 3]
```diff
@@ -75,10 +82,24 @@ int main(int argc, char *argv[]) {
...
+      std::cerr << "Could not parse table entries: "
+                << table_entries_or_status.status() << std::endl;
+      return 1;
+    }
+    table_entries = table_entries_or_status.value();
+  }
+
   // Transform to IR and print.
   pdpi::StatusOr<p4_symbolic::ir::P4Program> ir_status =
-      p4_symbolic::ir::Bmv2AndP4infoToIr(bmv2_or_status.value(),
```
Comment (konne88) [inline p4_symbolic/main.cc:80]: I also left this comment in the other CL:

create a Status Main() function that uses Status macros, and only do printing and return 1 once in the main

File: p4_symbolic/ir/table_entries.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#include <utility>
+#include <vector>
+
+#include "p4_pdpi/utils/status_utils.h"
+#include "p4_symbolic/ir/ir.pb.h"
+
+namespace p4_symbolic {
+namespace ir {
+
+using TableEntries = std::vector<std::pair<std::string, TableEntry>>;
```
Comment (konne88) [inline p4_symbolic/ir/table_entries.h]: add a comment, especially important, what is that std::string? the table name?

File: p4_symbolic/ir/ir.cc [hunk 5]
```diff
@@ -344,6 +346,36 @@ pdpi::StatusOr<P4Program> Bmv2AndP4infoToIr(const bmv2::P4Program &bmv2,
   }
   output.set_initial_table(bmv2.pipelines(0).init_table());
 
+  // Parse Table entries and fill them in the IR.
+  for (const std::pair<std::string, TableEntry> &pair : table_entries) {
+    const TableEntry &entry = pair.second;
```
Comment (konne88) [inline p4_symbolic/ir/ir.cc]: Use a struct instead of pair, so you can have good names.

File: p4_symbolic/ir/ir.cc [hunk 6]
```diff
@@ -344,6 +346,36 @@ pdpi::StatusOr<P4Program> Bmv2AndP4infoToIr(const bmv2::P4Program &bmv2,
...
+    // Table and action aliases parsed from table entris files
+    // will be replaced with their respective full name.
+    const std::string &table_alias = pair.first;
+    const std::string &action_alias = entry.action();
+
+    if (pdpi.tables_by_name().count(table_alias) != 1) {
+      return absl::Status(
+          absl::StatusCode::kInvalidArgument,
+          absl::StrCat("Table ", table_alias,
```
Comment (konne88) [inline p4_symbolic/ir/ir.cc]: RET_CHECK

File: p4_symbolic/ir/ir.cc [hunk 7]
```diff
@@ -344,6 +346,36 @@ pdpi::StatusOr<P4Program> Bmv2AndP4infoToIr(const bmv2::P4Program &bmv2,
...
+                       " from entries is missing from p4info!"));
+    }
+    if (pdpi.actions_by_name().count(action_alias) != 1) {
```
Comment (konne88) [inline p4_symbolic/ir/ir.cc]: RET_CHECK, so it doesn't distract from the main code so much

File: p4_symbolic/ir/ir.cc [hunk 8]
```diff
@@ -344,6 +346,36 @@ pdpi::StatusOr<P4Program> Bmv2AndP4infoToIr(const bmv2::P4Program &bmv2,
...
+          absl::StrCat("Action ", action_alias,
+                       " from entries is missing from p4info!"));
+    }
+
+    const std::string &table_name =
+        pdpi.tables_by_name().at(table_alias).preamble().name();
+    TableEntry *dst = (*output.mutable_tables())[table_name]
+                          .mutable_table_implementation()
+                          ->add_entries();
+    dst->CopyFrom(entry);
```
Comment (konne88) [inline p4_symbolic/ir/ir.cc]: *dst = entry;

**Review by konne88 — CHANGES_REQUESTED** [review]

File: p4_symbolic/main.cc (hunk 3 above)
Comment (KinanBab) [reply p4_symbolic/main.cc:80]: I am going to do that in the very last CL to avoid git rebasing conflict issues

**Comment by KinanBab** [comment]
Address most comments, blocked on RET_CHECK getting added to pdpi.

**Comment by KinanBab** [comment]
All comments have been address: main cleanup is on #30.

**Pull request merged**